### PR TITLE
cleanup(driver,userspace/libscap): let kmod use syscalls mask instead of events mask, like bpf and modern bpf

### PR DESCRIPTION
**Pull request by FedeDP**
**What type of PR is this?**

/kind cleanup

**Any specific area of the project related to this PR?**

/area driver-kmod
/area libscap-engine-kmod

**Does this PR require a change in the driver versions?**

/version driver-API-version-major

But we already bumped major in #709

**What this PR does / why we need it**:

Both bpf and modern bpf use a syscalls of interest concept; kmod was instead using an event of interest concept.
I ported kmod to use same concept of bpf and modern bpf.
NOTE: using event of interest also carried bugs, because when an user enforce certain `ppm_sc` through the scap/sinsp APIs, we would also lose all other non-syscalls event types generated by kmod (ie: signals, drops...), because kmod internally zeroed the event mask.

**Which issue(s) this PR fixes**:

Described above.

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
cleanup(driver)!: port kmod to use syscalls mask instead of events mask.
```

### CODE REVIEW

File: driver/main.c
Comment (Andreagit97) [inline driver/main.c:528]: Why do we need to enable all the syscalls here?

**Review by Andreagit97 — COMMENTED** [review]
Just one minor question

**Comment by gnosek** [comment]
This needs an API version bump, and ideally a new ioctl name/number instead of changing the behavior of existing ones

**Comment by Andreagit97** [comment]
I agree with the rename but do we really need the old feature on `events_mask`? do we have to disable events that are not syscalls?

**Comment by Andreagit97** [comment]
> We probably have to do this via tracepoints now.

The tracepoint events shouldn't be masked, they should be always available, we just directly attach and detach the tracepoints :thinking: if the tracepoint is not there it cannot generate any event

> In any case, we cannot just change the ioctl behavior and use the same number (there's tons of fun to be had if you e.g. use an ancient consumer with a new driver). IMO:
> 
>     introduce a completely new ioctl for this
>     make the old ioctl return -ENOTTY or something so that we can't mess up by accidentally mismatching versions

Absolutely agree with that!

File: driver/main.c
Comment (FedeDP) [reply driver/main.c:528]: Yep this is a default; i don't think it is necessary given that scap_kmod will set all the interesting syscalls at startup time.

**Comment by FedeDP** [comment]
> This needs an API version bump, and ideally a new ioctl name/number instead of changing the behavior of existing ones

Api version bump, as said in the op, was already done in this release cycle for #709; no more changes are required.
For the ioctl: note that an old client is not expected to work, given the api maj version bump; that is the reason we added those semver checks!

File: driver/main.c
Comment (FedeDP) [reply driver/main.c:528]: I'd leave it there for now :)

**Comment by FedeDP** [comment]
> I still disagree about this but I know I'm outnumbered so man_shrugging upside_down_face

We are going to reach chromium version in a couple of weeks if we bump every time we break :laughing:

**Comment by FedeDP** [comment]
@gnosek everything should be ok now :) I dropped the old ioctls and added new ones.

File: userspace/libscap/scap_vtable.h
Comment (gnosek) [inline userspace/libscap/scap_vtable.h:43]: I don't remember if we rely on these values to save a bit of work in kmod, but they're outdated now (they refer to the just-removed ioctls)

Not sure if we want to do something about it, your call

**Comment by dwindsor** [comment]
> Both bpf and modern bpf use a syscalls of interest concept; kmod was instead using an event of interest concept.
I ported kmod to use same concept of bpf and modern bpf.

Where does this leave the non-syscall kmod tracepoints (`sched_process_exit`, `page_fault_user` et al)?

**Comment by FedeDP** [comment]
> Where does this leave the non-syscall kmod tracepoints (sched_process_exit, page_fault_user et al)?

Well, the `ppm_sc` API is for disabling syscalls (while leaving `sys_enter` and `sys_exit` tracepoints attached). You can still disable tracepoints too.
Basically, we have events that maps 1:1 with tracepoints for all tracepoints except `sys_enter` and `sys_exit`.
These latter tracepoints map 1:N with events (ie: each tracepoints manages all syscall related enter/exit events).

**Comment by dwindsor** [comment]
Makes sense. I guess what I was getting at was [this](https://github.com/falcosecurity/libs/blob/6f9bec480954056233816e196753ea9c11b73da2/driver/ppm_events_public.h#L1804):

```c
// #define PPM_IOCTL_DISABLE_SIGNAL_DELIVER _IO(PPM_IOCTL_MAGIC, 14) Support dropped
// #define PPM_IOCTL_ENABLE_SIGNAL_DELIVER _IO(PPM_IOCTL_MAGIC, 15) Support dropped
```

☝️this change seems fairly recent, does it mean that we can no longer disable these tracepoints (there are others that were disabled as well).

**Comment by FedeDP** [comment]
Of course you can :) but there is now a proper Scap API for that and a TPMASK ioctl! (A bit lower in that very same file)

**Comment by FedeDP** [comment]
Answered above to grzeg with the same question! :smile:
See the OP of this PR

**Review by Andreagit97 — APPROVED** [review]
/approve

**Comment by FedeDP** [comment]
@gnosek mind to approve if you are ok?

**Review by hbrueckner — APPROVED** [review]
/lgtm

**Review by gnosek — APPROVED** [review]

**Pull request merged**
